# Beispiel: Mietwohnungen in Düsseldorf, aufgeschlüsselt nach Quartieren

Wir wollen jetzt an einem konkreten Beispiel einige Features von Pandas testen. Bisher haben wir mit "schönen" Daten gearbeitet, doch in der Realität hat man es leider nicht immer so leicht. Das folgende Beispiel wurde von der Website 

https://opendata.duesseldorf.de/dataset/wohnungen-den-wohnquartieren-von-d%C3%BCsseldorf

heruntergeladen. Hier ist es der Datensatz

__Wohnungen nach Baujahr Wohnquartiere Düsseldorf 2017__

Ein aktueller Datensatz liegt nicht vor.

Laut Website enthält jeder Datensatz die folgenden Daten (Spalten):

* Wohnquartier
* Anzahl der Wohnungen in Nichtwohngebäuden
* Anzahl der Wohnungen in Wohngebäuden insgesamt
* Wohnfläche je Wohnung
* Anzahl der Wohnungen in Wohngebäuden mit Baujahr vor 1949
* Anzahl der Wohnungen in Wohngebäuden mit Baujahr 1949 – 1969
* Anzahl der Wohnungen in Wohngebäuden mit Baujahr 1970 – 1989
* Anzahl der Wohnungen in Wohngebäuden mit Baujahr nach 1990
* Anzahl der Wohnungen in EZFH insgesamt
* Anzahl der Wohnungen in EZFH mit Baujahr vor 1949
* Anzahl der Wohnungen in EZFH mit Baujahr 1949 – 1969
* Anzahl der Wohnungen in EZFH mit Baujahr 1970 – 1989
* Anzahl der Wohnungen in EZFH mit Baujahr nach 1990
* Anzahl der Wohnungen in MFH insgesamt
* Anzahl der Wohnungen in MFH mit Baujahr vor 1949
* Anzahl der Wohnungen in MFH mit Baujahr 1949 – 1969
* Anzahl der Wohnungen in MFH mit Baujahr 1970 – 1989
* Anzahl der Wohnungen in MFH mit Baujahr nach 1990
* Durchschnittliche Anzahl Wohnungen je Wohngebäude

In [1]:
import pandas as pd

### Datei einlesen

Wir können uns die Datei `wohnungen.csv` direkt ansehen. Wir erkennen, dass die Wohnungsdaten wie beschrieben nach `wohnquartier` aufgeschlüsselt sind. Außerdem sind die Werte mit Semikola getrennt, und unsere Float-Werte verwenden ein Komma statt eines Dezimalpunktes.

Wir lesen die Datei also wie folgt ein:

In [2]:
wohnungen = pd.read_csv("wohnungen.csv", delimiter=';',  decimal=",",  index_col = "wohnquartier");
wohnungen

bestand_in_nichtwohngebaeuden  \
wohnquartier                                  
11001                                 452.0   
11002                                   NaN   
12001                                 393.0   
12002                                   NaN   
13001                                   NaN   
...                                     ...   
101014                                  NaN   
102001                                  NaN   
102002                                  NaN   
102003                                  4.0   
102004                                  4.0   

              bestand_in_wohngebaeuden_insgesamt  wohnflaeche_je_wohnung  \
wohnquartier                                                               
11001                                     1485.0                   67.06   
11002                                        NaN                     NaN   
12001                                     1407.0                   79.42   
12002                                        NaN                     NaN   
13001                                        NaN                     NaN   
...                                          ...                     ...   
101014                                       NaN                     NaN   
102001                                     476.0                  106.01   
102002                                    1081.0                   94.47   
102003                                    1081.0                  103.10   
102004                                       6.0                   97.33   

              bestand_in_wohngebaeuden_mit_bj_vor_1949  \
wohnquartier                                             
11001                                            258.0   
11002                                              NaN   
12001                                            454.0   
12002                                              NaN   
13001                                              NaN   
...                                                ...   
101014                                             NaN   
102001                                             3.0   
102002                                            21.0   
102003                                             8.0   
102004                                             0.0   

              bestand_in_wohngebaeuden_mit_bj_1949_bis_1969  \
wohnquartier                                                  
11001                                                 563.0   
11002                                                   NaN   
12001                                                 472.0   
12002                                                   NaN   
13001                                                   NaN   
...                                                     ...   
101014                                                  NaN   
102001                                                  NaN   
102002                                                 15.0   
102003                                                  NaN   
102004                                                  0.0   

              bestand_in_wohngebaeuden_mit_bj_1970_bis_1989  \
wohnquartier                                                  
11001                                                 401.0   
11002                                                   NaN   
12001                                                 342.0   
12002                                                   NaN   
13001                                                   NaN   
...                                                     ...   
101014                                                  NaN   
102001                                                467.0   
102002                                                271.0   
102003                                                965.0   
102004                                                  3.0   

              best

# Anzeige der Metadaten

Wir verschaffen uns einen kleinen Überblick, indem wir die Metadaten anzeigen lassen:

In [3]:
print(wohnungen.info())

<class 'pandas.core.frame.DataFrame'>
Index: 373 entries, 11001 to 102004
Data columns (total 18 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   bestand_in_nichtwohngebaeuden                  267 non-null    float64
 1   bestand_in_wohngebaeuden_insgesamt             319 non-null    float64
 2   wohnflaeche_je_wohnung                         319 non-null    float64
 3   bestand_in_wohngebaeuden_mit_bj_vor_1949       318 non-null    float64
 4   bestand_in_wohngebaeuden_mit_bj_1949_bis_1969  316 non-null    float64
 5   bestand_in_wohngebaeuden_mit_bj_1970_bis_1989  317 non-null    float64
 6   bestand_in_wohngebaeuden_mit_bj_nach_1990      317 non-null    float64
 7   bestand_in_ezfh_insgesamt                      297 non-null    float64
 8   bestand_in_ezfh_mit_bj_vor_1949                216 non-null    float64
 9   bestand_in_ezfh_mit_bj_1949_bis_1969           215 n

Wir sehen:

1. Die Bezeichnung der Spalten entspricht der obigen Beschreibung. Durch die Schreibweise mit dem _Underscore_ können wir auch über den Spaltennamen auf die Spalte zugreifen
3. Es sind 373 Datensätze vorhanden.
5. Bis auf den Wert für `wohnquartier` (der Index) sind für keinen Wert alle Daten vollständig vorhanden.



## Zugriff auf einzelne Daten

### Zugriff auf Spalten

Wir können eine einzelne Spalte extrahieren, indem wir als Index den Spaltennamen verwenden. Das Ergbnis ist eine `Series`:


In [4]:
bestand_in_mfh = wohnungen['bestand_in_in_mfh_insgesamt' ]
print(f'Daten-Typ bestand_in_mfh: {type(bestand_in_mfh)}')
print(bestand_in_mfh)
q = 102003
print(f'\nBestand in Quartier {q}: {bestand_in_mfh[q]:0.0f}')

Daten-Typ bestand_in_mfh: <class 'pandas.core.series.Series'>
wohnquartier
11001     1415.0
11002        NaN
12001     1351.0
12002        NaN
13001        NaN
           ...  
101014       NaN
102001     199.0
102002     591.0
102003     410.0
102004       3.0
Name: bestand_in_in_mfh_insgesamt, Length: 373, dtype: float64

Bestand in Quartier 102003: 410


### Summen über Spalten

Wir können numerische Werte einfach summieren, wie in `NumPy. Fehlende Werte (`NaN`) werden dabei ignoriert:

In [5]:
bestand_mfh_ges = int( bestand_in_mfh.sum())
print(f'Gesamtbestand: {bestand_mfh_ges:,}'.replace(",","."))

Gesamtbestand: 299.782


### Einfache Statistiken

Ein wichtiges Feature ist `describe()`. Hier wird eine einfache Statistik über sämtliche numerische Spalten erzeugt und angezeigt:

In [20]:
wohnungen.describe()

bestand_in_nichtwohngebaeuden  bestand_in_wohngebaeuden_insgesamt  \
count                     267.000000                          319.000000   
mean                       52.074906                         1072.012539   
std                        63.154557                          398.473344   
min                         3.000000                            6.000000   
25%                        13.000000                          822.000000   
50%                        33.000000                         1076.000000   
75%                        67.000000                         1328.000000   
max                       452.000000                         2357.000000   

       wohnflaeche_je_wohnung  bestand_in_wohngebaeuden_mit_bj_vor_1949  \
count              319.000000                                318.000000   
mean                76.642602                                232.150943   
std                 17.242261                                206.834751   
min                 20.250000                                  0.000000   
25%                 64.575000                                 77.500000   
50%                 71.700000                                182.000000   
75%                 83.620000                                329.000000   
max                148.220000                               1066.000000   

       bestand_in_wohngebaeuden_mit_bj_1949_bis_1969  \
count                                     316.000000   
mean                                      471.221519   
std                                       305.898585   
min                                         0.000000   
25%                                       231.750000   
50%                                       438.500000   
75%                                       673.750000   
max                                      1428.000000   

       bestand_in_wohngebaeuden_mit_bj_1970_bis_1989  \
count                                     317.000000   
mean                                      216.381703   
std                                       187.184137   
min                                         0.000000   
25%                                        89.000000   
50%                                       173.000000   
75%                                       271.000000   
max                                       989.000000   

       bestand_in_wohngebaeuden_mit_bj_nach_1990  bestand_in_ezfh_insgesamt  \
count                                 317.000000                 297.000000   
mean                                  159.744479                 141.932660   
std                                   176.059364                 175.615815   
min                                     0.000000                   0.000000   
25%                                    43.000000                  12.000000   
50%                                   112.000000                  72.000000   
75%                                   216.000000                 206.000000   
max                                  1694.000000                 912.000000   

       bestand_in_ezfh_mit_bj_vor_1949  bestand_in_ezfh_mit_bj_1949_bis_1969  \
count                       216.000000                            215.000000   
mean                         52.504630                             62.437209   
std                          69.139629                             72.600523   
min                           3.000000                              3.000000   
25%                           8.000000                             10.000000   
50%                          27.000000                             32.000000   
75%                          67.250000                             90.000000   
max                         516.000000                            399.000000   

       bestand_in_ezfh_mit_bj_1970_bis_1989  bestand_in_ezfh_mit_bj_nach_1990  \
count                            175.000000                        184.000000   
mean                              48.1542

## Fehlende Daten

Wir haben gesehen, dass der Datensatz nicht vollständig ist. In allen Spalten fehlen Daten, und es gibt sogar einzeln Zeilen, die komplett leer sind. Hier wurde einige Quartiere nicht erfasst.

Wie man fachlich mit  fehlenden Daten umgeht, ist Sache der Statistiker (oder Data Scientists). Technisch bietet `Pandas` alle Möglichkeiten, Datensätze entsprechend zu manipulieren.

### Leere Zeilen entfernen

Als erstes löschen wir aus unserem Datensatz alle leeren Zeilen. Zum Entfernen verwenden wir die Funktion `dropna()

In [6]:
wohnungen1 = wohnungen.dropna()
wohnungen1

bestand_in_nichtwohngebaeuden  \
wohnquartier                                  
11001                                 452.0   
12001                                 393.0   
15003                                   7.0   
16002                                  52.0   
23005                                   8.0   
...                                     ...   
96003                                  42.0   
96004                                  29.0   
96005                                   4.0   
98006                                  55.0   
98009                                  16.0   

              bestand_in_wohngebaeuden_insgesamt  wohnflaeche_je_wohnung  \
wohnquartier                                                               
11001                                     1485.0                   67.06   
12001                                     1407.0                   79.42   
15003                                      753.0                   71.83   
16002                                      658.0                  125.06   
23005                                     2357.0                   81.73   
...                                          ...                     ...   
96003                                     1038.0                   88.78   
96004                                     1125.0                   79.74   
96005                                      826.0                  108.26   
98006                                     1196.0                   74.39   
98009                                     1323.0                   81.86   

              bestand_in_wohngebaeuden_mit_bj_vor_1949  \
wohnquartier                                             
11001                                            258.0   
12001                                            454.0   
15003                                            424.0   
16002                                            223.0   
23005                                            484.0   
...                                                ...   
96003                                            287.0   
96004                                            198.0   
96005                                            122.0   
98006                                            139.0   
98009                                            289.0   

              bestand_in_wohngebaeuden_mit_bj_1949_bis_1969  \
wohnquartier                                                  
11001                                                 563.0   
12001                                                 472.0   
15003                                                 255.0   
16002                                                 138.0   
23005                                                1428.0   
...                                                     ...   
96003                                                 258.0   
96004                                                 423.0   
96005                                                 362.0   
98006                                                 396.0   
98009                                                 456.0   

              bestand_in_wohngebaeuden_mit_bj_1970_bis_1989  \
wohnquartier                                                  
11001                                                 401.0   
12001                                                 342.0   
15003                                                  52.0   
16002                                                  53.0   
23005                                                 223.0   
...                                                     ...   
96003                                                 313.0   
96004                                                 378.0   
96005                                                 209.0   
98006                                                 303.0   
98009                                                 164.0   

              best

Hoppla! jetzt sind nur noch 107 Zeilen von 373 übrig. So schlecht war die Datenlage aber gar nicht.

`dropna()` löscht alle Zeilen, in denen _ein_ Datensatz fehlt. Wir wollten aber nur komplett leere Zeilen entfernen. Das geht soe:

In [7]:
wohnungen1 = wohnungen.dropna(how="all")
wohnungen1

bestand_in_nichtwohngebaeuden  \
wohnquartier                                  
11001                                 452.0   
12001                                 393.0   
13002                                 430.0   
13003                                 274.0   
13004                                  82.0   
...                                     ...   
101012                                  NaN   
102001                                  NaN   
102002                                  NaN   
102003                                  4.0   
102004                                  4.0   

              bestand_in_wohngebaeuden_insgesamt  wohnflaeche_je_wohnung  \
wohnquartier                                                               
11001                                     1485.0                   67.06   
12001                                     1407.0                   79.42   
13002                                     1073.0                   65.44   
13003                                     1678.0                   61.71   
13004                                     1171.0                   60.21   
...                                          ...                     ...   
101012                                     164.0                  114.99   
102001                                     476.0                  106.01   
102002                                    1081.0                   94.47   
102003                                    1081.0                  103.10   
102004                                       6.0                   97.33   

              bestand_in_wohngebaeuden_mit_bj_vor_1949  \
wohnquartier                                             
11001                                            258.0   
12001                                            454.0   
13002                                            114.0   
13003                                             92.0   
13004                                             89.0   
...                                                ...   
101012                                            19.0   
102001                                             3.0   
102002                                            21.0   
102003                                             8.0   
102004                                             0.0   

              bestand_in_wohngebaeuden_mit_bj_1949_bis_1969  \
wohnquartier                                                  
11001                                                 563.0   
12001                                                 472.0   
13002                                                 816.0   
13003                                                1365.0   
13004                                                 940.0   
...                                                     ...   
101012                                                 12.0   
102001                                                  NaN   
102002                                                 15.0   
102003                                                  NaN   
102004                                                  0.0   

              bestand_in_wohngebaeuden_mit_bj_1970_bis_1989  \
wohnquartier                                                  
11001                                                 401.0   
12001                                                 342.0   
13002                                                  50.0   
13003                                                 153.0   
13004                                                 115.0   
...                                                     ...   
101012                                                 38.0   
102001                                                467.0   
102002                                                271.0   
102003                                                965.0   
102004                                                  3.0   

              best

Schon besser. Nur etwa 60 Zeilen wurden entfernt. Wie sieht die Lage nun aus?

In [8]:
wohnungen1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 319 entries, 11001 to 102004
Data columns (total 18 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   bestand_in_nichtwohngebaeuden                  267 non-null    float64
 1   bestand_in_wohngebaeuden_insgesamt             319 non-null    float64
 2   wohnflaeche_je_wohnung                         319 non-null    float64
 3   bestand_in_wohngebaeuden_mit_bj_vor_1949       318 non-null    float64
 4   bestand_in_wohngebaeuden_mit_bj_1949_bis_1969  316 non-null    float64
 5   bestand_in_wohngebaeuden_mit_bj_1970_bis_1989  317 non-null    float64
 6   bestand_in_wohngebaeuden_mit_bj_nach_1990      317 non-null    float64
 7   bestand_in_ezfh_insgesamt                      297 non-null    float64
 8   bestand_in_ezfh_mit_bj_vor_1949                216 non-null    float64
 9   bestand_in_ezfh_mit_bj_1949_bis_1969           215 n

Nun ist das Verhältnis fehlender Daten zu allen Daten nicht mehr so dramatisch. Einige Spalten sind sogar komplett gefüllt.

### Fehlende Werte auffüllen

Will man mit allen Zeilen und Spalten weiterarbeiten, so ist man i.a. gezwungen, die fehlenden Werte zu ergänzen, da sonst einige Berechnungen fehlschlagen werden. Das einfachste ist es, fehlenden Werte durch eine feste Zahl zu ergänzen:


In [13]:
wohnungen2 = wohnungen1.fillna(0)
wohnungen2

bestand_in_nichtwohngebaeuden  \
wohnquartier                                  
11001                                 452.0   
12001                                 393.0   
13002                                 430.0   
13003                                 274.0   
13004                                  82.0   
...                                     ...   
101012                                  0.0   
102001                                  0.0   
102002                                  0.0   
102003                                  4.0   
102004                                  4.0   

              bestand_in_wohngebaeuden_insgesamt  wohnflaeche_je_wohnung  \
wohnquartier                                                               
11001                                     1485.0                   67.06   
12001                                     1407.0                   79.42   
13002                                     1073.0                   65.44   
13003                                     1678.0                   61.71   
13004                                     1171.0                   60.21   
...                                          ...                     ...   
101012                                     164.0                  114.99   
102001                                     476.0                  106.01   
102002                                    1081.0                   94.47   
102003                                    1081.0                  103.10   
102004                                       6.0                   97.33   

              bestand_in_wohngebaeuden_mit_bj_vor_1949  \
wohnquartier                                             
11001                                            258.0   
12001                                            454.0   
13002                                            114.0   
13003                                             92.0   
13004                                             89.0   
...                                                ...   
101012                                            19.0   
102001                                             3.0   
102002                                            21.0   
102003                                             8.0   
102004                                             0.0   

              bestand_in_wohngebaeuden_mit_bj_1949_bis_1969  \
wohnquartier                                                  
11001                                                 563.0   
12001                                                 472.0   
13002                                                 816.0   
13003                                                1365.0   
13004                                                 940.0   
...                                                     ...   
101012                                                 12.0   
102001                                                  0.0   
102002                                                 15.0   
102003                                                  0.0   
102004                                                  0.0   

              bestand_in_wohngebaeuden_mit_bj_1970_bis_1989  \
wohnquartier                                                  
11001                                                 401.0   
12001                                                 342.0   
13002                                                  50.0   
13003                                                 153.0   
13004                                                 115.0   
...                                                     ...   
101012                                                 38.0   
102001                                                467.0   
102002                                                271.0   
102003                                                965.0   
102004                                                  3.0   

              best

In [12]:
wohnungen2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 319 entries, 11001 to 102004
Data columns (total 18 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   bestand_in_nichtwohngebaeuden                  319 non-null    float64
 1   bestand_in_wohngebaeuden_insgesamt             319 non-null    float64
 2   wohnflaeche_je_wohnung                         319 non-null    float64
 3   bestand_in_wohngebaeuden_mit_bj_vor_1949       319 non-null    float64
 4   bestand_in_wohngebaeuden_mit_bj_1949_bis_1969  319 non-null    float64
 5   bestand_in_wohngebaeuden_mit_bj_1970_bis_1989  319 non-null    float64
 6   bestand_in_wohngebaeuden_mit_bj_nach_1990      319 non-null    float64
 7   bestand_in_ezfh_insgesamt                      319 non-null    float64
 8   bestand_in_ezfh_mit_bj_vor_1949                319 non-null    float64
 9   bestand_in_ezfh_mit_bj_1949_bis_1969           319 n

Jetzt sind alle Werte gefüllt. Aber das ist natürlich eher _brute force_. Eleganter wäre es etwa, in jeder Spalte fehlende Werte durch den Mittelwert über diese Spalte aufzufüllen. Dann sieht für die ertse Spalte so aus:


In [18]:
wohnungen2['bestand_in_nichtwohngebaeuden'] = wohnungen1['bestand_in_nichtwohngebaeuden'].fillna(wohnungen1['bestand_in_nichtwohngebaeuden'].mean())
wohnungen2

bestand_in_nichtwohngebaeuden  \
wohnquartier                                  
11001                            452.000000   
12001                            393.000000   
13002                            430.000000   
13003                            274.000000   
13004                             82.000000   
...                                     ...   
101012                            52.074906   
102001                            52.074906   
102002                            52.074906   
102003                             4.000000   
102004                             4.000000   

              bestand_in_wohngebaeuden_insgesamt  wohnflaeche_je_wohnung  \
wohnquartier                                                               
11001                                     1485.0                   67.06   
12001                                     1407.0                   79.42   
13002                                     1073.0                   65.44   
13003                                     1678.0                   61.71   
13004                                     1171.0                   60.21   
...                                          ...                     ...   
101012                                     164.0                  114.99   
102001                                     476.0                  106.01   
102002                                    1081.0                   94.47   
102003                                    1081.0                  103.10   
102004                                       6.0                   97.33   

              bestand_in_wohngebaeuden_mit_bj_vor_1949  \
wohnquartier                                             
11001                                            258.0   
12001                                            454.0   
13002                                            114.0   
13003                                             92.0   
13004                                             89.0   
...                                                ...   
101012                                            19.0   
102001                                             3.0   
102002                                            21.0   
102003                                             8.0   
102004                                             0.0   

              bestand_in_wohngebaeuden_mit_bj_1949_bis_1969  \
wohnquartier                                                  
11001                                                 563.0   
12001                                                 472.0   
13002                                                 816.0   
13003                                                1365.0   
13004                                                 940.0   
...                                                     ...   
101012                                                 12.0   
102001                                                  0.0   
102002                                                 15.0   
102003                                                  0.0   
102004                                                  0.0   

              bestand_in_wohngebaeuden_mit_bj_1970_bis_1989  \
wohnquartier                                                  
11001                                                 401.0   
12001                                                 342.0   
13002                                                  50.0   
13003                                                 153.0   
13004                                                 115.0   
...                                                     ...   
101012                                                 38.0   
102001                                                467.0   
102002                                                271.0   
102003                                                965.0   
102004                                                  3.0   

              best

# Zusammenfassung

Bevor Daten mit Künstlicher Intelligenz verarbeitet werden können, müssen sie entsprechend aufbereitet werden. Erhobene Daten, insbesondere im soziologischen bereich, sind meist unvollständig. Die Statistik kennt entsprechende Methoden, mit solchen Daten umzugehen (z.B. durch _Imputation_). 

`Pandas` bietet alle erdenklichen Möglichkeiten, auch mit großen Datenmengen umzugehen. Dazu gehören

* Speichen, Laden und Konvertieren von Daten
* Zugriff auf und Anzeige von Datensätzen
* Manipulation einzelner Zellen, Zeilen, Spalten und kompletter Datensätze
* Statistische Berechnungen

Dabei lieg `Pandas` das numerische `Numpy`-Paket zugrunde. Wer `Numpy`beherrscht, wird sich mit `Pandas` nicht schwer tun.t Zusammn mit `Numpy`und `Matplotlib` bildet `Pandas`den grundlegenden Werkzeugkasten für _Data Scientists_, die mit Python arbeiten wollen.
